x/ecocredit/server/basket/put_test.go

```diff
@@ -0,0 +1,353 @@
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/golang/mock/gomock"
+	basketv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1"
+	"github.com/regen-network/regen-ledger/orm"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	basket2 "github.com/regen-network/regen-ledger/x/ecocredit/basket"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/basket"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server/basket/mocks"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"testing"
+	"time"
+)
+
+func TestPut(t *testing.T) {
+	basketDenom := "BASKET"
+	basketDenom2 := "BASKET2"
+	classId := "CARB2"
+	startDate, err := time.Parse("2006-01-02", "2020-01-01")
+	require.NoError(t, err)
+	endDate, err := time.Parse("2006-01-02", "2021-01-01")
+	require.NoError(t, err)
+	denom, err := ecocredit.FormatDenom(classId, 1, &startDate, &endDate)
+	require.NoError(t, err)
+	testClassInfo := ecocredit.ClassInfo{
+		ClassId:    classId,
+		Admin:      "somebody",
+		Issuers:    nil,
+		Metadata:   nil,
+		CreditType: &ecocredit.CreditType{
+			Name:         "carbon",
+			Abbreviation: "C01",
+			Unit:         "many carbons",
+			Precision:    6,
+		},
+		NumBatches: 1,
+	}
+	classInfoRes := ecocredit.QueryClassInfoResponse{Info: &testClassInfo}
+	testBatchInfo := ecocredit.BatchInfo{
+		ClassId:         classId,
+		BatchDenom:      denom,
+		Issuer:          "somebody",
+		TotalAmount:     "1000000000000000000000000000",
+		Metadata:        nil,
+		AmountCancelled: "0",
+		StartDate:       &startDate,
+		EndDate:         &endDate,
+		ProjectLocation: "US-NY",
+	}
+	batchInfoRes := ecocredit.QueryBatchInfoResponse{Info: &testBatchInfo}
+
+
+	ctx := context.Background()
+	 _,_, addr := testdata.KeyTestPubAddr()
+	ctrl := gomock.NewController(t)
+	b := ormtest.NewMemoryBackend()
+	db, err := ormdb.NewModuleDB(server.ModuleSchema, ormdb.ModuleDBOptions{
+		GetBackend: func(ctx context.Context) (ormtable.Backend, error) {
+			return b, nil
+		},
+		GetReadBackend: func(ctx context.Context) (ormtable.ReadBackend, error) {
+			return b, nil
+		},
+	})
+	require.NoError(t, err)
+	basketTbl := db.GetTable(&basketv1.Basket{})
+	err = basketTbl.Insert(ctx, &basketv1.Basket{
+		BasketDenom:       basketDenom,
+		DisableAutoRetire: true,
+		CreditTypeName:    "carbon",
+		DateCriteria:     &basketv1.DateCriteria{Sum: &basketv1.DateCriteria_MinStartDate{MinStartDate: timestamppb.New(startDate)}},
+		Exponent:          6,
+	})
+	require.NoError(t, err)
+	var dur time.Duration = -500000000000000000
```

Comment:
```suggestion
	var dur time.Duration = 500000000000000000
```

pretty sure this should be positive (I had another comment to switch `Add` to `Sub` )